experimental/cuda2/pending_queue_actions.c

```diff
@@ -525,20 +692,128 @@ iree_status_t iree_hal_cuda2_pending_queue_actions_issue(
-    action->event_count = 0;
-
-    action = next_action;
+  iree_hal_cuda2_atomic_slist_entry_t* entry = NULL;
+  // TODO: avoid host allocator malloc; use some pool for the allocation.
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0, iree_allocator_malloc(actions->host_allocator, sizeof(*entry),
+                                (void**)&entry));
+  entry->ready_list_head = ready_list.head;
+
+  // Now push the ready list to the worker and have it to issue the actions to
+  // the GPU.
+  iree_hal_cuda2_working_area_t* working_area = &actions->working_area;
+  iree_hal_cuda2_ready_action_slist_push(&working_area->ready_worklist, entry);
+  iree_hal_cuda2_worker_state_t prev_state =
+      (iree_hal_cuda2_worker_state_t)iree_atomic_exchange_int32(
+          &working_area->worker_state,
+          IREE_HAL_CUDA_WORKER_STATE_WORKLOAD_PENDING,
+          iree_memory_order_acq_rel);
+  iree_notification_post(&working_area->state_notification, IREE_ALL_WAITERS);
+
+  // Handle potential error cases from the worker thread.
+  iree_status_t status = iree_ok_status();
+  if (prev_state == IREE_HAL_CUDA_WORKER_STATE_EXIT_ERROR) {
+    iree_status_code_t code = iree_atomic_load_int32(&working_area->error_code,
+                                                     iree_memory_order_acquire);
+    status = iree_status_from_code(code);
   }
 
   IREE_TRACE_ZONE_END(z0);
+  return status;
+}
+
+//===----------------------------------------------------------------------===//
+// Worker routines
+//===----------------------------------------------------------------------===//
+
+static bool iree_hal_cuda2_worker_has_incoming_request(
+    iree_hal_cuda2_working_area_t* working_area) {
+  iree_hal_cuda2_worker_state_t value = iree_atomic_load_int32(
+      &working_area->worker_state, iree_memory_order_acquire);
+  return value == IREE_HAL_CUDA_WORKER_STATE_WORKLOAD_PENDING ||
+         value == IREE_HAL_CUDA_WORKER_STATE_EXIT_REQUESTED;
+}
+
+static bool iree_hal_cuda2_worker_committed_exiting(
+    iree_hal_cuda2_working_area_t* working_area) {
+  return iree_atomic_load_int32(&working_area->worker_state,
+                                iree_memory_order_acquire) ==
+         IREE_HAL_CUDA_WORKER_STATE_EXIT_COMMITTED;
+}
+
+// Processes all ready actions in the given |worklist|.
+static iree_status_t iree_hal_cuda2_worker_process_ready_list(
+    iree_allocator_t host_allocator,
+    iree_hal_cuda2_ready_action_slist_t* worklist) {
+  iree_hal_cuda2_atomic_slist_entry_t* entry =
+      iree_hal_cuda2_ready_action_slist_pop(worklist);
+  while (entry) {
+    // Process the current batch of ready actions.
+    for (iree_hal_cuda2_queue_action_t* action = entry->ready_list_head;
+         action != NULL;) {
+      iree_hal_cuda2_queue_action_t* next_action = action->next;
+      action->next = NULL;
+
+      iree_status_t status =
+          iree_hal_cuda2_pending_queue_actions_issue_execution(action);
+      if (!iree_status_is_ok(status)) return status;
```

Comment:
I think this early exit will leak the entry - tracking multiple return sites gets really hard to read/modify and we should try to reduce the number (there's a lot in this new code).

if you rotate the loop so it's a do-while with the first pop on the inside it gets easier to follow:
```
iree_status_t status = iree_ok_status();
do {
  entry = pop(worklist);
  for (...) {
    status = issue_execution(action);
    if (!iree_status_is_ok(status)) break;
  }
  free(entry);
} while (iree_status_is_ok(status));
```

Reply:
Ah good catch! Yes, indeed.